### PR TITLE
LM-40: Flawed validation logic for already booked events allows booking multiple events

### PR DESCRIPTION
**Pull request by kkuczkowski-unosq**
Jira issue: # [LM-40](https://unosquarebelfast.atlassian.net/browse/LM-40)

- Fixed predicate logic to disallow booking multiple events on a single day across already booked event
- Adjusted existing and added new unit tests to add test coverage for the code changes

### CODE REVIEW

File: AdminCore.Services/EventService.cs [hunk 1]
```diff
@@ -311,9 +316,21 @@ private bool IsDateRangeLessThanTotalHolidaysRemaining(EventDateDto eventDates,
...
-        GetApprovedEventDatesByEmployeeAndStartAndEndDates(eventDates.StartDate, eventDates.EndDate, employeeId);
-      if (employeeEvents.Any())
+      var approvedEmployeeEvents = GetApprovedEventDatesByEmployeeAndStartAndEndDatesAndStatus(
+        eventDates.StartDate,
+        eventDates.EndDate,
+        employeeId,
+        EventStatuses.Approved
+        );
+
+      var awaitEmployeeEvents = GetApprovedEventDatesByEmployeeAndStartAndEndDatesAndStatus(
```
Comment (rgillen-uno) [inline AdminCore.Services/EventService.cs]: Maybe call 'GetApprovedEventDatesByEmployeeAndStartAndEndDatesAndStatus' something more generic? Maybe 'GetEventDatesByEmployeeAndStartAndEndDatesAndStatus'

**Review by rgillen-uno — APPROVED** [review]
Worked through the new logic, looks good to me.

File: AdminCore.Services/EventService.cs [hunk 2]
```diff
@@ -53,15 +53,20 @@ public IList<EventDto> GetEventsByEmployeeId(int employeeId, EventTypes eventTyp
...
-                                                                         && x.EndDate.Date <= endDate.Date
-                                                                         || x.EndDate.Date == startDate.Date)
-                                                                         && x.Event.EmployeeId == employeeId
-                                                                         && x.Event.EventStatusId == eventStatus,
-                                                              null, x => x.Event);
+    public IList<EventDateDto> GetEventDatesByEmployeeAndStartAndEndDatesAndStatus(DateTime startDate, DateTime endDate, int employeeId, EventStatuses eventStatuses)
+    {
+      var eventStatus = (int)eventStatuses;
+      var eventDates = DatabaseContext.EventDatesRepository.Get(
+        current =>
```
Comment (rgillen-uno) [inline AdminCore.Services/EventService.cs]: Just had a thought, should tests be written to cover all date scenarios to ensure that logic isn't missed?

```
Date        5   6   7   8
Booked  :   o           o
New     :   o       o

Booked  :   o           o
New     :       o   o

Booked  :   o           o
New     :       o       o

Booked  :   o           o
New     :   o           o

Booked  :       o   o
New     :   o           o
```

Comment (BrownMark) [reply AdminCore.Services/EventService.cs]: Makes sense

Comment (BrownMark) [reply AdminCore.Services/EventService.cs]: We could also update the IRepository for another method to check if something exists. 

Something like: TestRepository.Exists(x => x.id == 1) 
which returns a boolean

**Review by rgillen-uno — APPROVED** [review]
New Tests look good. Covers all date possibilities.

**Pull request merged**